Chunk SingleNodeShutdownStatus and ShutdownShardMigrationStatus (and related action) response

ldematte opened the pull request:
Related to https://github.com/elastic/elasticsearch/issues/97803 and others.

ShardAllocationDecision grows with the number of nodes. Therefore, it would save heap space to chunk ShutdownShardMigrationStatus as well.

At this point, to close the circle, it makes sense to also make `SingleNodeShutdownStatus` (which depends on `ShutdownShardMigrationStatus`) chunked, and modify accordingly `GetShutdownStatusAction.Response` and `RestGetShutdownStatusAction` to return a chunked response.

Closes https://github.com/elastic/elasticsearch/issues/99678

ldematte committed:
Moving SingleNodeShutdownStatus and ShutdownShardMigrationStatus to chunked response

ldematte committed:
Update docs/changelog/99798.yaml

elasticsearchmachine commented:
Hi @ldematte, I've created a changelog YAML for you.

elasticsearchmachine commented:
Pinging @elastic/es-core-infra (Team:Core/Infra)

DaveCTurner commented on server/src/main/java/org/elasticsearch/common/xcontent/ChunkedToXContentHelper.java:
```diff
@@ -74,6 +74,20 @@ public static Iterator<ToXContent> field(String name, String value) {
+     * @param value ChunkedToXContent value for this field (single value, object or array)
+     * @return Iterator composing field name and value serialization
+     */
+    public static Iterator<ToXContent> field(String name, ChunkedToXContent value) {
```
I think this has to be a full object, not a fragment:

```suggestion
    public static Iterator<ToXContent> field(String name, ChunkedToXContentObject value) {
```

ldematte replied:
Yes, good one, I tried to "enforce" this via the javadoc comment but using the more specific type is much better

DaveCTurner commented on server/src/main/java/org/elasticsearch/common/xcontent/ChunkedToXContentHelper.java:
```diff
@@ -74,6 +74,20 @@ public static Iterator<ToXContent> field(String name, String value) {
+    public static Iterator<ToXContent> field(String name, ChunkedToXContent value) {
+        return Iterators.concat(
+            Iterators.single((builder, params) -> builder.field(name)),
+            value.toXContentChunked(ToXContent.EMPTY_PARAMS)
```
I think we need to propagate the params from the caller rather than using `EMPTY_PARAMS` here. It might well not matter right now but it will catch us out in future.

ldematte replied:
If you think it's safer, I will change it. 
The reason why I used `EMPTY_PARAMS` is that the original code calls into this `XContentBuilder` [function](https://github.com/elastic/elasticsearch/blob/2df6def54fbff76a1d0108cbef7d9dca46757741/libs/x-content/src/main/java/org/elasticsearch/xcontent/XContentBuilder.java#L963), which calls [this function](https://github.com/elastic/elasticsearch/blob/2df6def54fbff76a1d0108cbef7d9dca46757741/libs/x-content/src/main/java/org/elasticsearch/xcontent/XContentBuilder.java#L971) with `EMPTY_PARAMS`. 
Should I provide both overloads, or maybe just the one with `params` to encourage people to propagate them?

DaveCTurner replied:
I'd prefer just the one with params so that people have to propagate. The one without params in `XContentBuilder` is equally trappy IMO.

ldematte replied:
Makes sense, done!

DaveCTurner reviewed (commented):
Looks good, just a couple of small comments about the new helper function.

ldematte committed:
Addressed PR comments

- Propagating Params to inner toXContentChunked call
- Using ChunkedToXContentObject to enforce non-fragment values

ldematte committed:
Merge branch 'chunk-shutdown-shard-migration-status-response' of github.com:ldematte/elasticsearch into chunk-shutdown-shard-migration-status-response

ldematte committed:
Merge remote-tracking branch 'upstream/main' into chunk-shutdown-shard-migration-status-response

DaveCTurner reviewed (approved):
LGTM
